documentation/architecture.md

```diff
@@ -0,0 +1,336 @@
+Instead of `kill`'ing the `cc-shim` PID, it will go through the following steps:
+
+1. `cc-oci-runtime` connects to `cc-proxy` and sends it the `attach` command to let it know
+on which pod the container it tries to `kill` is running.
+2. `cc-oci-runtime` sends an hyperstart `KILLCONTAINER` command to `kill` the container running
+on the guest. The command is sent to `cc-proxy` who forwards it to the right hyperstart instance
+running in the appropriate guest.
+
+##### `delete`
+
+`docker delete` is about deleting all resources held by a stopped/killed container. Running
+containers can not be `delete`d unless the OCI runtime is explictly being asked to. In that
+case it will first `kill` the container and only then `delete` it.
+
+The resources held by a Clear Container are quite different from the ones held by a host
+namespace container e.g. run by `runc`. `cc-oci-runtime` needs mostly to delete the pod
+holding the stopped container on the virtual machine, shut the hypervisor down and finally
+delete all related proxy resources:
+
+1. `cc-oci-runtime` connects to `cc-proxy` and sends it the `attach` command to let it know
+on which pod the container it tries to `delete` is running.
+2. `cc-oci-runtime` sends an hyperstart `DESTROYPOD` command to `destroy` the pod holding the
+container running on the guest. The command is sent to `cc-proxy` who forwards it to the right
+hyperstart instance running in the appropriate guest.
+3. After deleting the last running pod, `hyperstart` will gracefully shut the virtual machine
+down.
+4. `cc-oci-runtime` sends the `BYE` command to `cc-proxy`, to let it know that a given virtual
+machine shut down. `cc-proxy` will then clean all its internal resources associated with this
+VM.
+
+#### Proxy
+
+`cc-proxy` is a daemon offering access to the [`hyperstart`](https://github.com/hyperhq/hyperstart)
+VM agent to multiple clients like for example `cc-shim` and `cc-proxy`.
+Only a single instance of `cc-proxy` per host is necessary as it can be used for several different VMs.
+Its main role is to:
+- Arbitrate access to the `hyperstart` control channel between all the `cc-oci-runtime` instances and the `cc-shim` ones.
+- Route the I/O streams between the various `cc-shim` instances and `hyperstart`.
+
+`cc-proxy` provides 2 client interfaces:
+
+- A UNIX, named socket for all `cc-oci-runtime` instances on the host to send commands to `cc-proxy`.
+- One socket pair per `cc-shim` instance, to send stdin and receive stdout and stderr I/O streams. See the
+[cc-shim section](https://github.com/01org/cc-oci-runtime/blob/topic/documentation/documentation/architecture.md#shim)
+for more details about that interface.
+
+The protocol on the `cc-proxy` UNIX named socket supports the following commands:
+
+- `Hello`: This command is for `cc-oci-runtime` to let `cc-proxy` know about a newly created VM that will
+hold containers. This command payload contains the `hyperstart` QEMU exported control and I/O UNIX socket
```

Comment:
I don't understand the phrase 'hyperstart QEMU exported control....' - any way we can clarify this?